tests/dopamine/utils/threading_utils_test.py

```diff
@@ -0,0 +1,224 @@
+
+  def testDefaultValueIsUsed(self):
+    """Tests that the default value is properly set by the helper."""
+    obj = _DummyClass()
+    threading_utils.initialize_local_attributes(
+        obj, attr=lambda: 'default-value')
+    self.assertEqual(obj.attr, 'default-value')
+
+  def testMultipleDefaultValuesAreUsed(self):
+    """Tests that multiple default values are properly set by the helper."""
+    obj = _DummyClass()
+    threading_utils.initialize_local_attributes(
+        obj, attr1=lambda: 3, attr2=lambda: 4)
+    self.assertEqual(obj.attr1, 3)
+    self.assertEqual(obj.attr2, 4)
+
+  def testMultiThreads(self):
+    """Tests that different threads create different local attributes."""
+    MockClass = threading_utils.local_attributes(['attr'])(_DummyClass)
+    obj = MockClass()
+    # Initializes attribute in thread 1.
+    with test_utils.mock_thread('thread_1'):
+      obj.attr = 1
+    # Initializes attribute in thread 2.
+    with test_utils.mock_thread('thread_2'):
+      obj.attr = 2
+    # Reads attribute in thread 1.
+    with test_utils.mock_thread('thread_1'):
+      self.assertEqual(obj.attr, 1)
+    # Reads attribute in thread 2.
+    with test_utils.mock_thread('thread_2'):
+      self.assertEqual(obj.attr, 2)
+
+  def testMultiThreadsMultipleAttributes(self):
+    """Tests that different threads create different local attributes."""
+    MockClass = threading_utils.local_attributes(
+        ['attr1', 'attr2'])(_DummyClass)
+    obj = MockClass()
+    # Initializes attribute in thread 1.
+    with test_utils.mock_thread('thread_1'):
+      obj.attr1 = 1
+      obj.attr2 = 2
+    with test_utils.mock_thread('thread_2'):
+      obj.attr1 = 3
+      obj.attr2 = 4
+    with test_utils.mock_thread('thread_1'):
+      self.assertEqual(obj.attr1, 1)
+      self.assertEqual(obj.attr2, 2)
+    with test_utils.mock_thread('thread_2'):
+      self.assertEqual(obj.attr1, 3)
+      self.assertEqual(obj.attr2, 4)
+
+  def testCallableAttribute(self):
+    """Tests that different attributes are initialized in each thread.."""
+    MockClass = threading_utils.local_attributes(['attr'])(_DummyClass)
+    obj = MockClass()
+    threading_utils.initialize_local_attributes(
+        obj, attr=lambda: test.mock.Mock())  # pylint: disable=unnecessary-lambda
+    # Call attribute in thread 1.
+    with test_utils.mock_thread('thread_1'):
+      obj.attr()
+    # Call attribute in thread 2.
+    with test_utils.mock_thread('thread_2'):
+      obj.attr()
+
+    # Check that attribute was called once in thread 1.
+    with test_utils.mock_thread('thread_1'):
+      obj.attr.assert_called_once()
+
+    # Check that attribute was called once in thread 2.
+    with test_utils.mock_thread('thread_2'):
+      obj.attr.assert_called_once()
+
+    # Check that attribute was not called in thread 3.
+    with test_utils.mock_thread('thread_3'):
+      obj.attr.assert_not_called()
+
+
+class ThreadingUtilsImplementationTest(test.TestCase):
```

Comment:
Continuing the discussion at https://github.com/wes-turner/dopamine/pull/4#discussion_r255312605:

> I don't mean to push back, I am open to alternatives. What do you think?

Pushing back (both when reviewer and reviewee) is one of my favorite parts of code reviews.  Those are the times I look most closely at my assumptions and am most likely to learn something.  Please keep doing so!

> that means I have no assurance that this is using the wrapped setter / getter over the default one.

Mmm.  That does seem like an important thing to be sure of.  Maybe we can come up with a scheme together that does that without examining the implementation.  With that assurance, I think the class above is the only one this test file needs.  Maybe since all the tests are on `Mock`s, is this guarantee already built-in?

https://go/tott/299 ("Test Behavior, Not Implementation") and https://go/tott/358 ("Prefer Testing Public APIs Over Implementation-Detail Classes") are both relevant.  358 says:

> Testing implementation-detail classes is still useful in many cases, such as if the class is complex or if the tests would be difficult to write for the public API class. 

I'm just not sure if that's the case, here.

In any event and no matter what you choose, this is ok to submit.  I understand the desire for additional safety, and if this gets in the way of future development, the tests are easy enough to remove -- especially when packaged this way, with the test class written above.

Reply:
Actually I was able to convert most of the tests to use only the API.

If tests are done within a `_mock_thread` then they make sense.

Reply:
I think the tests look good.  Nice work.